### PR TITLE
batches: handle colon in search query

### LINKED ISSUE #38737: Create batch changes flow breaks when query contains unescaped characters
#### Current state

Creating a batch change from this doesn't "just work":
https://demo.sourcegraph.com/search?q=context:global+fork:yes+repo:%5Egithub%5C.com/yowayb/uBlock%24+file:package.json+%22scaling-palm-tree%22:+%22...%22&patternType=structural

Error: Failed to normalize JSON: yaml: line 6: mapping values are not allowed in this context

#### Impact

https://sourcegraph.slack.com/archives/CMMTWQQ49/p1657717252405149

#### Desired state

Quote the search string to avoid this

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes #38737

When a colon is included in a search query (example: `context:global fork:yes repo:^github\.com/foo/bar$ file:package.json "@types/node": "..." patternType:structural`), creating a batch change from said query results in an error in the `Configuration` page. This happens because the YAML parser assumes the colon is signifying a node.

An easy fix for this is to leverage `multi-line` values in YAML so the parser understands that it's part of the current node.

I took the approach of checking if appending the search query to `- repositoriesMatchingQuery` results in a valid YAML, if it doesn't then I make it a multi-line YAML.

## Test plan

Tested with search query containing an extra colon

## App preview:

- [Web](https://sg-web-bo-fix-batch-change-from-search.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-bybpsavhsz.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by courier-new — APPROVED** [review]
This is great! Do you think we could add a test case to `yaml-util` for this edge case, and maybe throw in a comment about why we're doing this extra check for posterity?

**Comment by BolajiOlajide** [comment]
> there is `quoteYAMLString` already in this file, can we make use of that instead?

Quoting it doesn't fix the problem. I tried using the `quoteYAMLString` method and still got the same error. 
Simply quoting the string didn't escape the colon character, I tried that approach initially and it didn't work. @eseliger

**Comment by eseliger** [comment]
I think we've hit an edge case here because we don't yet properly escape `"` in the quoted string, but the helper and quoting in general should work 🤔 can we add that fix to the helper and then use it?

**Comment by BolajiOlajide** [comment]
> I think we've hit an edge case here because we don't yet properly escape `"` in the quoted string, but the helper and quoting in general should work 🤔 can we add that fix to the helper and then use it?

The indentation is pretty specific to the `- repositoriesMatchingQuery: ` line in the batch spec.
I'm happy to try out another approach, but using the helper did not fix the problem. Here's a screenshot of the YAML when I try to escape the `"`.

CleanShot 2022-07-14 at 00 00 26@2x

What I noticed is that escaping the `"` isn't the problem, the main problem is escaping the `:`, the `quoteYAMLString` doesn't give that.
When I tried to quote the colon manually, here's what the query looked like and I'm not wondering if this isn't a bit confusing to the user - also because it's a structural search, the workspace resolver might not return the correct match since it'll search for a quoted colon also.

```yml
on:
  - repositoriesMatchingQuery: context:global fork:yes repo:^github\.com/yowayb/uBlock$ file:package.json "scaling-palm-tree"'':'' "..." patternType:structural
```

**Comment by eseliger** [comment]
you didn't escape the last two `"`, using `"context:global repo:github.com/foo/bar file:package.json fork:yes \"@types:node\": \"...\" patternType:structural"` works well for me, with all `"` replaced by `\"`.

**Comment by courier-new** [comment]
_yelling in yaml_

**Comment by BolajiOlajide** [comment]
> "context:global repo:github.com/foo/bar file:package.json fork:yes \"@types:node\": \"...\" patternType:structural"

Oh yeah! Thanks for catching that @eseliger  - this pattern works when the query doesn't contain a regex, however when it does, it doesn't. You can try with the original query from Yoway here: `fork:yes repo:^github\.com/yowayb/uBlock$ file:package.json "scaling-palm-tree": "..."`

CleanShot 2022-07-14 at 11 05 05@2x

**Comment by BolajiOlajide** [comment]
> Would this help: https://go.dev/play/p/sNpBtVHZbT4?

Well, the problem with this is that the resolver isn't able to parse the regex properly, therefore we are unable to resolve the workspaces.

CleanShot 2022-07-14 at 12 16 40@2x

**Comment by BolajiOlajide** [comment]
> Sorry, I may be missing something here completely, but encoding things when quoting a string should be fully transparent. It would all be reversed when getting the string value for the potentially quoted field. In fact, I tried it locally and `"github\\.com\\/sourcegraph"` when passed to the backend perfectly is just `github\.com\/sourcegraph` as expected and the search returns the expected results. What am I missing? :D

Honestly, I'm confused as to why we have two different results. I also assumed encoding should be transparent but I just don't get that on my end. Let's spend a few minutes to take a look at this during our meeting today.

File: client/web/src/enterprise/batches/batch-spec/yaml-util.ts [hunk 1]
```diff
@@ -347,38 +348,58 @@ export const isMinimalBatchSpec = (spec: string): boolean => {
 /**
  * Inspects a given string value and determines if it needs to be quoted to produce valid yaml.
  *
+ * @param key the key of the string value to be inspected
  * @param value the string value to inspect
  */
-export function quoteYAMLString(value: string): string {
+export function quoteYAMLString(key: string, value: string): string {
```
Comment (eseliger) [inline client/web/src/enterprise/batches/batch-spec/yaml-util.ts:354]: I think the key being `name` here might've been a bit misleading but for yaml quoting the `key`s value doesn't actually matter afaik. It could've been anything in this function, I just chose name arbitrarily. Could also have used `testValue:`. Is there any reason I'm missing? Otherwise we can remove the `key` param and keep this function very simple to use.

**Review by eseliger — APPROVED** [review]

File: client/web/src/enterprise/batches/batch-spec/yaml-util.ts (hunk 1 above)
Comment (BolajiOlajide) [reply client/web/src/enterprise/batches/batch-spec/yaml-util.ts]: Naaah, you're right. The key doesn't matter atleast at this present moment.
I've made the change.
